fdbserver/KeyValueStoreShardedRocksDB.actor.cpp

```diff
@@ -489,22 +502,153 @@ struct PhysicalShard {
+				break;
+			}
+			cursor->Next();
+		}
+		s = cursor->status();
+		shard->physicalShard->readIterPool->returnIterator(readIter);
+	} else {
+		ReadIterator readIter = shard->physicalShard->readIterPool->getIterator();
+		auto cursor = readIter.iter;
+		cursor->SeekForPrev(toSlice(range.end));
+		if (cursor->Valid() && toStringRef(cursor->key()) == range.end) {
+			cursor->Prev();
+		}
+		while (cursor->Valid() && toStringRef(cursor->key()) >= range.begin) {
+			KeyValueRef kv(toStringRef(cursor->key()), toStringRef(cursor->value()));
+			++accumulatedRows;
+			accumulatedBytes += sizeof(KeyValueRef) + kv.expectedSize();
+			result->push_back_deep(result->arena(), kv);
+			// Calling `cursor->Prev()` is potentially expensive, so short-circut here just in case.
+			if (result->size() >= -rowLimit || accumulatedBytes >= byteLimit) {
+				break;
+			}
+			cursor->Prev();
+		}
+		s = cursor->status();
+		shard->physicalShard->readIterPool->returnIterator(readIter);
+	}
+
+	if (!s.ok()) {
+		logRocksDBError(s, "ReadRange");
+		// The data writen to the arena is not erased, which will leave RangeResult in a dirty state. The RangeResult
+		// should never be returned to user.
+		return -1;
+	}
+	return accumulatedBytes;
+}
+
 // Manages physical shards and maintains logical shard mapping.
 class ShardManager {
 public:
 	ShardManager(std::string path) : path(path) {}
 	rocksdb::Status init() {
-		// TODO: Open default cf and get shard mapping.
+		dataShardMap.insert(allKeys, nullptr);
 		// Open instance.
-		auto options = getOptions();
-		rocksdb::Status status = rocksdb::DB::Open(options, path, &db);
+		std::vector<std::string> columnFamilies;
+		rocksdb::Options options = getOptions();
+		rocksdb::Status status = rocksdb::DB::ListColumnFamilies(options, path, &columnFamilies);
+
+		rocksdb::ColumnFamilyOptions cfOptions = getCFOptions();
+		std::vector<rocksdb::ColumnFamilyDescriptor> descriptors;
+		bool foundMetadata = false;
+		for (const auto& name : columnFamilies) {
+			if (name == "kvs-metadata") {
+				foundMetadata = true;
+			}
+			descriptors.push_back(rocksdb::ColumnFamilyDescriptor{ name, cfOptions });
+		}
+
+		// Add default column family if it's a newly opened database.
+		if (descriptors.size() == 0) {
+			descriptors.push_back(rocksdb::ColumnFamilyDescriptor{ "default", cfOptions });
+		}
+
+		std::vector<rocksdb::ColumnFamilyHandle*> handles;
+		status = rocksdb::DB::Open(options, path, descriptors, &handles, &db);
 		if (!status.ok()) {
 			logRocksDBError(status, "Open");
 			return status;
 		}
-		metadataShard = std::make_shared<PhysicalShard>(db, "kvs-metadata");
+
+		if (foundMetadata) {
+			for (auto handle : handles) {
+				if (handle->GetName() == "kvs-metadata") {
+					metadataShard = std::make_shared<PhysicalShard>(db, "kvs-metadata", handle);
+				} else {
+					physicalShards[handle->GetName()] = std::make_shared<PhysicalShard>(db, handle->GetName(), handle);
```

Comment:
Since we are doing initialization, we should expect no conflicts, i.e., the following should hold?
```
auto [it, inserted] = physicalShards.emplace(handle->GetName(), std::make_shared<PhysicalShard>(db, handle->GetName(), handle));

ASSERT(inserted);
```